File: crates/kernel/chio-process/examples/workers/inventory.py
"""Deterministic source inventory worker; no model or network credentials."""

import json
import sys

from chio_process import ProcessClient

config = json.load(sys.stdin)
client = ProcessClient(config["socket_path"], config["credential"])
read = client.invoke("read-snapshot", "tools", "read", {})
if read["verdict"] != "allow":
    raise RuntimeError("source snapshot denied")
files = read["output"]["value"]["files"]
report = {
    "worker": "python",
    "files": len(files),
    "nonempty_lines": sum(
        bool(line.strip()) for file in files for line in file["content"].split("\n")
    ),
    "paths": sorted(file["path"] for file in files),
}
published = client.invoke("publish-inventory", "tools", "append", report)
if published["verdict"] != "allow":
    raise RuntimeError("inventory publication denied")
snapshot = client.inspect()
if snapshot["checkpoint"]["revision"] == "0":
    blob = client.put_blob(bytes(range(256)) * 4096)
    client.checkpoint("0", {"published": True, "blob": blob})
reference = client.inspect()["checkpoint"]["value"]["blob"]
if client.read_blob(reference["sha256"]) != bytes(range(256)) * 4096:
    raise RuntimeError("persisted blob differs")
print(json.dumps({"read": read, "published": published, "snapshot": client.inspect()}))

File: crates/kernel/chio-process/examples/workers/inventory.mjs
import { readFileSync } from "node:fs";
import { ProcessClient } from "../../../../../sdks/typescript/packages/process/index.mjs";

// Deterministic source inventory worker; no model or network credentials.
const config = JSON.parse(readFileSync(0, "utf8"));
const client = new ProcessClient(config.socket_path, config.credential);
const read = await client.invoke("read-snapshot", "tools", "read", {});
if (read.verdict !== "allow") throw new Error("source snapshot denied");
const files = read.output.value.files;
const report = {
  worker: "javascript",
  files: files.length,
  nonempty_lines: files.reduce((sum, file) => sum + file.content.split("\n").filter(line => line.trim()).length, 0),
  paths: files.map(file => file.path).sort(),
};
const published = await client.invoke("publish-inventory", "tools", "append", report);
if (published.verdict !== "allow") throw new Error("inventory publication denied");
const snapshot = await client.inspect();
const bytes = Uint8Array.from({ length: 1_048_576 }, (_, index) => index % 256);
if (snapshot.checkpoint.revision === "0") {
  const blob = await client.putBlob(bytes);
  await client.checkpoint("0", { published: true, blob });
}
const reference = (await client.inspect()).checkpoint.value.blob;
const stored = await client.readBlob(reference.sha256);
if (stored.length !== bytes.length || stored.some((byte, index) => byte !== bytes[index])) throw new Error("persisted blob differs");
console.log(JSON.stringify({ read, published, snapshot: await client.inspect() }));
